sdk/include/FreeRTOS-Compat/semphr.h

```diff
@@ -53,34 +178,272 @@ xSemaphoreCreateCounting(UBaseType_t uxMaxCount, UBaseType_t uxInitialCount)
 }
 
 /**
- * Delete a heap-allocated semaphore.
+ * Create a heap-allocated binary semaphore.
+ *
+ * Binary semaphores are implemented as counting semaphores with a maximum count
+ * of 1.
+ */
+__always_inline static inline SemaphoreHandle_t xSemaphoreCreateBinary()
+{
+	return xSemaphoreCreateCounting(1, 0);
+}
+
+/**
+ * Create a statically allocated binary semaphore.
+ *
+ * Binary semaphores are implemented as counting semaphores with a maximum count
+ * of 1.
+ */
+__always_inline static inline SemaphoreHandle_t
+xSemaphoreCreateBinaryStatic(StaticSemaphore_t *pxSemaphoreBuffer)
+{
+	return xSemaphoreCreateCountingStatic(1, 0, pxSemaphoreBuffer);
+}
+#endif
+
+#if CHERIOT_FREERTOS_MUTEX
+/**
+ * Create statically allocated mutex.
+ */
+__always_inline static inline SemaphoreHandle_t
+xSemaphoreCreateMutexStatic(StaticSemaphore_t *pxMutexBuffer)
+{
+	pxMutexBuffer->mutex.lockWord = 0;
+	CHERIOT_FREERTOS_SEMAPHORE_KIND_SET(pxMutexBuffer, CheriotFreeRTOS_Mutex);
+	return pxMutexBuffer;
+}
+
+/**
+ * Create a heap-allocated mutex.
+ */
+__always_inline static inline SemaphoreHandle_t xSemaphoreCreateMutex()
+{
+	SemaphoreHandle_t semaphore =
+	  (SemaphoreHandle_t)malloc(sizeof(SemaphoreHandle_t));
+	if (semaphore != NULL)
+	{
+		xSemaphoreCreateMutexStatic(semaphore);
+	}
+	return semaphore;
+}
+#endif
+
+#if CHERIOT_FREERTOS_RECURSIVE_MUTEX
+
+/**
+ * Create a statically allocated recursive mutex.
+ */
+__always_inline static inline SemaphoreHandle_t
+xSemaphoreCreateRecursiveMutexStatic(StaticSemaphore_t *pxMutexBuffer)
+{
+	pxMutexBuffer->recursiveMutex.lock.lockWord = 0;
+	pxMutexBuffer->recursiveMutex.depth         = 0;
+	CHERIOT_FREERTOS_SEMAPHORE_KIND_SET(pxMutexBuffer,
+	                                    CheriotFreeRTOS_RecursiveMutex);
+	return pxMutexBuffer;
+}
+
+/**
+ * Create a heap-allocated recursive mutex.
+ */
+__always_inline static inline SemaphoreHandle_t
+xSemaphoreCreateRecursiveMutex(void)
+{
+	SemaphoreHandle_t semaphore =
+	  (SemaphoreHandle_t)malloc(sizeof(SemaphoreHandle_t));
+	if (semaphore != NULL)
+	{
+		xSemaphoreCreateRecursiveMutexStatic(semaphore);
+	}
+	return semaphore;
+}
+#endif
+
+/**
+ * Delete a heap-allocated semaphore, of any kind.
+ *
+ * Note: As on FreeRTOS, if there are waiters blocked on this semaphore then
+ * they will remain blocked until their timeout (if ever).
  */
 __always_inline static inline void
 vSemaphoreDelete(SemaphoreHandle_t xSemaphore)
 {
 	free(xSemaphore);
 }
 
+/**
+ * Get the current owner (thread ID) of a mutex.  If this is a counting
+ * semaphore, or if the lock is not held, returns zero.
```

Comment:
This probably deserves the same comment about racing as below, with the (anti)caveat that an answer of "this thread" _is_ stable.